### PR TITLE
Add disable module check for storage APIs

### PR DESCRIPTION
**Pull request by anjbur**
Things to consider before you submit the PR:

* [ ] Has `CHANGELOG.md` been updated?
* [x] Are tests passing locally?
* [x] Are the files formatted correctly?
* [x] Did you add unit tests?
* [ ] Did you check UI tests on the sample app? They are not executed on CI.
* [ ] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?

## Description

Add checks and tests in API for storage module being disabled.

## Related PRs or issues

[AB#59476](https://msmobilecenter.visualstudio.com/web/wi.aspx?pcguid=27bf89de-9697-418a-accf-1aa125b22914&id=59476)

### CODE REVIEW

File: AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m [hunk 1]
```diff
@@ -300,25 +317,36 @@ - (void)deleteDocumentWithPartition:(NSString *)partition
...
-                     }
-                     completionHandler([[MSDataSourceError alloc] initWithError:cosmosDbError]);
-                   }];
+
+  @synchronized(self) {
+    if (![self canBeUsed] || ![self isEnabled]) {
+      NSError *error = [[NSError alloc] initWithDomain:kMSACErrorDomain
+                                                  code:MSACDisabledErrorCode
+                                              userInfo:@{NSLocalizedDescriptionKey : kMSACDisabledErrorDesc}];
+      MSLogError([MSDataStore logTag], @"Not able to delete the document ID:%@ with error:%@", documentId, [error description]);
```
Comment (achocron) [inline AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m]: Can the NSError and log be extracted into a method? So the method will return the `NSError` and the methods will just invoke the completion handler after that.

Comment (achocron) [reply AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m]: Seems like it would be a bit more flexible

Comment (anjbur) [reply AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m]: Fair, refactored that helper method.

File: AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m [hunk 2]
```diff
@@ -234,53 +235,68 @@ - (void)replaceWithPartition:(NSString *)partition
                    completionHandler:completionHandler];
 }
 
+- (NSError *)generateDisabledError:(NSString *)operation documentId:(NSString *)documentId {
+  NSError *error = [[NSError alloc] initWithDomain:kMSACErrorDomain
+                                              code:MSACDisabledErrorCode
+                                          userInfo:@{NSLocalizedDescriptionKey : kMSACDisabledErrorDesc}];
+  MSLogError([MSDataStore logTag], @"Not able to %@ the document ID:%@ with error:%@", operation, documentId, [error description]);
```
Comment (jaelim-ms) [inline AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m]: Use `localizedDescription`.

File: AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m [hunk 3]
```diff
@@ -300,25 +316,33 @@ - (void)deleteDocumentWithPartition:(NSString *)partition
...
-                     } else {
-                       MSLogDebug([MSDataStore logTag], @"Document deleted: %@/%@", partition, documentId);
-                     }
-                     completionHandler([[MSDataSourceError alloc] initWithError:cosmosDbError]);
-                   }];
+
+  @synchronized(self) {
+    if (![self canBeUsed] || ![self isEnabled]) {
+      NSError *error = [self generateDisabledError:@"delete" documentId:documentId];
+      completionHandler([[MSDataSourceError alloc] initWithError:error errorCode:MSACDocumentUnknownErrorCode]);
```
Comment (jaelim-ms) [inline AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m]: Why is the code `MSACDocumentUnknownErrorCode`? I wonder why it takes `errorCode` along with `error` instance.

File: AppCenterDataStorage/AppCenterDataStorageTests/MSDataStoreTests.m [hunk 4]
```diff
@@ -363,6 +514,7 @@ - (void)testCreateWithPartitionGoldenPath {
       });
 
   // When
+  XCTAssertTrue([[MSDataStore sharedInstance] canBeUsed]);
```
Comment (jaelim-ms) [inline AppCenterDataStorage/AppCenterDataStorageTests/MSDataStoreTests.m]: This is your test code. Please remove.

File: AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m (hunk 3 above)
Comment (jwallra) [reply AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m]: I guess my question is why it's not `documentId` like it is for `readWithPartition`

File: AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m (hunk 2 above)
Comment (jwallra) [inline AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m]:
```suggestion
  MSLogError([MSDataStore logTag], @"Not able to %@ the document ID: %@; error: %@", operation, documentId, [error description]);
```

File: AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m [hunk 5]
```diff
@@ -328,154 +352,175 @@ - (void)createOrReplaceWithPartition:(NSString *)partition
...
-                       MSPaginatedDocuments *documents = [[MSPaginatedDocuments alloc] initWithError:dataSourceDeserializeError];
-                       completionHandler(documents);
-                       return;
-                     }
+  @synchronized(self) {
+    if (![self canBeUsed] || ![self isEnabled]) {
+      NSError *error = [[NSError alloc] initWithDomain:kMSACErrorDomain
+                                                  code:MSACDisabledErrorCode
+                                              userInfo:@{NSLocalizedDescriptionKey : kMSACDisabledErrorDesc}];
+      MSLogError([MSDataStore logTag], @"Not able to list the documents in partition:%@ with error:%@", partition, [error description]);
```
Comment (jwallra) [inline AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m]:
```suggestion
      MSLogError([MSDataStore logTag], @"Not able to list the documents in partition: %@; error: %@", partition, [error description]);
```

File: AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m (hunk 3 above)
Comment (guperrot) [reply AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m]: @jaelim-ms This errorCode expects to be from an enum that defines only http errors. The NSError has its own code for disabled.

Comment (guperrot) [reply AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m]: > I guess my question is why it's not `documentId` like it is for `readWithPartition`

We are using documentId here too, I don't understand the feedback.

Comment (guperrot) [reply AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m]: Also its different completion handlers for different APIs it's unrelated to our changes.

File: AppCenterDataStorage/AppCenterDataStorageTests/MSDataStoreTests.m [hunk 6]
```diff
@@ -171,6 +173,155 @@ - (void)testApplyEnabledStateWorks {
   XCTAssertEqual(enabledCount, 1);
 }
 
+- (void)testReadWithPartitionDisabled {
```
Comment (dhei) [inline AppCenterDataStorage/AppCenterDataStorageTests/MSDataStoreTests.m]: What is "PartitionDisabled"? Should it be "testReadWhenDataModuleDisabled"?

If you rename this test, please rename other tests that you added below.

Comment (dhei) [reply AppCenterDataStorage/AppCenterDataStorageTests/MSDataStoreTests.m]: Addressed myself.

**Pull request merged**